### PR TITLE
[feat] engine: support for Odysee

### PR DESCRIPTION
**Pull request by Austin-Olacsi**
## What does this PR do?

Adds the Odysee engine!

## Why is this change important?

People will like it, it was actually requested over on the searx github: https://github.com/searx/searx/issues/2504

## How to test this PR locally?

make a search like- test !od

## Author's checklist

the reason i also made odysee_videos.html is because the embedded videos did not look quite right when using videos.html

### CODE REVIEW

**Comment by Bnyro** [comment]
For reference here: #590

**Comment by Bnyro** [comment]
And please use the `videos.html` template (if someone else confirms that), I don't think it's worth adding a new template solely for odysee here.

**Comment by return42** [comment]
Hi @Hackurei thanks a lot for your PR .. before we go into a review, the patches should pass the quality gate from CI .. as you can see in "Some checks were not successful" there are some fixes needed to pass the quality gate .. I recommend to study at least our [Development Quickstart](https://docs.searxng.org/dev/quickstart.html) .. further it would be good to squash your 5 commits into one commit [[ref]](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes) with a [meaningful commit messages](https://docs.searxng.org/dev/contribution_guide.html#code).

**Comment by Austin-Olacsi** [comment]
hello, please understand i do not have time to work on this right now, i will come back in a few weeks / months.

**Comment by Bnyro** [comment]
The changes seem pretty decent so far, would you mind if anyone else continues working on your PR and finalizing it?

**Comment by Austin-Olacsi** [comment]
@Bnyro you can if you want. i still need to learn how to squash everything into 1 commit though. also embedded videos are looking good with the videos.html now, yay!

File: searx/engines/odysee.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    "official_api_documentation": None,
+    "use_official_api": False,
+    "require_api_key": False,
+    "results": "JSON",
+}
+
+# Engine configuration
+paging = True
+results_per_page = 20
+categories = ['images']
```
Comment (Bnyro) [inline searx/engines/odysee.py:24]:
```suggestion
categories = ['videos']
```

File: searx/engines/odysee.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+# Search URL (Note: lighthouse.lbry.com/search works too, and may be faster at times)
+base_url = "https://lighthouse.odysee.tv/search"
+
+# Request function
+def request(query, params):
+    page = params.get("pageno", 1) - 1
```
Comment (Bnyro) [inline searx/engines/odysee.py:31]: `params["pageno"]` is always defined afaik, so we don't have to use `Dict#get`?

---
update from @return42: fixed in https://github.com/searxng/searxng/commit/4f8895c6de2659f9ca069dde7be8eb3714ed1494

File: docs/dev/engines/online/odysee.rst [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+Odysee
+======
+
+.. contents:: Contents
+   :depth: 2
+   :local:
+   :backlinks: entry
+
+.. automodule:: searx.engines.odysee
+  :members:
```
Comment (return42) [inline docs/dev/engines/online/odysee.rst:13]: I added the framework for the documentation / generated from the doc-strings in the source code.

File: searx/engines/odysee.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    query_params = {
+        "s": query,
+        "size": results_per_page,
+        "from": start_index,
+        "include": "channel,thumbnail_url,title,description,duration,release_time",
+        "mediaType": "video",
+    }
+
+    query_str = "&".join([f"{key}={value}" for key, value in query_params.items()])
+    params["url"] = f"{base_url}?{query_str}"
```
Comment (Bnyro) [inline searx/engines/odysee.py:43]:
```suggestion
    params["url"] = f"{base_url}?{urlencode(query_params)}"
```

--- 
update from @return42: fixed in https://github.com/searxng/searxng/commit/4f8895c6de2659f9ca069dde7be8eb3714ed1494

File: searx/engines/odysee.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    page = params.get("pageno", 1) - 1
+    start_index = page * results_per_page
+
+    query_params = {
+        "s": query,
+        "size": results_per_page,
+        "from": start_index,
+        "include": "channel,thumbnail_url,title,description,duration,release_time",
+        "mediaType": "video",
+    }
```
Comment (return42) [inline searx/engines/odysee.py:40]: From where did you get this REST API? .. I looked at the [odysee-frontend](https://github.com/OdyseeTeam/odysee-frontend) .. but I just could not find an API description in a hurry .. if you have more information, let us know and we can add it to the documentation.

File: searx/engines/odysee.py (hunk 1 above)
Comment (return42) [inline searx/engines/odysee.py:24]: I removed the timeout from the settings.yml .. the default timeout has always been enough / I did not had any issues with timeouts. And I moved the definition of the categories from settings.yml file to here in the odysee.py module.

File: searx/engines/odysee.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+# lint: pylint
```
Comment (return42) [inline searx/engines/odysee.py:2]: I added the [license tag](https://github.com/searxng/searxng/blob/master/LICENSE) .. the `# lint: pylint` instructs our CI to perform a pylint with a more rigorous assessment --> all new implementation must withstand these rigorous assessments.

**Review by return42 — APPROVED** [review]
> i still need to learn how to squash everything into 1 commit though.

No problem .. I squashed the commits rebased your branch on SXNG's master and did some modifications (see my comments from the review).

> also embedded videos are looking good with the videos.html now, yay!

Works like a charme :) .. well done :+1: .. thanks a lot for your work to add odysee to SXNG / I think Odysee is a very valuable addition to the videos category ... maybe we can get more out of it ... I'd still like to find the API for it.

File: searx/engines/odysee.py (hunk 1 above)
Comment (return42) [reply searx/engines/odysee.py:24]: Argh .. I haven't seen your review comments before I merge this PR / sorry .. I will fix it in a follow up.

Comment (return42) [reply searx/engines/odysee.py:24]: @Bnyro thanks a lot for your review comments / I would like to apologize again for overlooking this in my haste / I added your review in commit https://github.com/searxng/searxng/commit/4f8895c6de2659f9ca069dde7be8eb3714ed1494 on the master branch.

Comment (Bnyro) [reply searx/engines/odysee.py:24]: No worries, thanks for doing the suggested cleanups!

File: searx/engines/odysee.py (hunk 5 above)
Comment (Austin-Olacsi) [reply searx/engines/odysee.py:40]: the only info i was able to find on this was posted by searx users here: https://github.com/searx/searx/issues/2504

based on this, it should be possible to add support for language filtering as well as safe search.

Comment (return42) [reply searx/engines/odysee.py:40]: > based on this, it should be possible to add support for language filtering as well as safe search.

would you like to add a [feature-request](https://github.com/searxng/searxng/issues/new?assignees=&labels=enhancement&projects=&template=feature-request.md&title=) .. just to have it on our *ToDo* list and not to forget / thanks!

**Pull request merged**